Add support for online payment of fines.

EreMaijala opened the pull request:
Adds online payment framework and payment gateway plugins for Stripe and Paytrail.

Also adds payment support to the following ILS drivers: Demo, KohaRest, SierraRest.

TODO:
- [x] Update change log when merged.
- [x] Add a [wiki page for online payment](https://vufind.org/wiki/configuration:online_payment).
- [x] Update ILS driver wiki page to include new methods and fine fields.
- [ ] Resolve [VUFIND-1518](https://openlibraryfoundation.atlassian.net/browse/VUFIND-1518) when done.

EreMaijala committed:
Add support for online payment of fines.

Adds online payment framework and payment gateway plugins for Stripe and Paytrail.

Also adds payment support to the following ILS drivers: Demo, KohaRest, SierraRest.

EreMaijala commented:
Note: While there is the test gateway for simulating a payment locally, Stripe's and Paytrail's test credentials also work well for more authentic experience:
- https://docs.stripe.com/testing
- https://docs.paytrail.com/#/test

EreMaijala committed:
Merge remote-tracking branch 'origin/dev' into dev-online-payment-support2

demiankatz commented on composer.json:
```diff
@@ -101,12 +102,14 @@
         "symfony/rate-limiter": "^6.4",
         "symfony/var-dumper": "6.4.21",
         "symfony/yaml": "6.4.21",
+        "tecnickcom/tcpdf": "6.10.0",
```
As you may already know, this library seems to be in the process of getting deprecated in favor of https://github.com/tecnickcom/tc-lib-pdf -- since the new one is not stable yet, I'm not sure there's much we can do about it right now, but something we'll need to keep an eye on.

EreMaijala replied:
I'm aware of that. But I can take a look, perhaps it's mature enough for our needs. Actually, looks like 8.1.0 is supposed to be stable, and it's now 8.1.5, so the README may be outdated.

EreMaijala replied:
I did not have any success with tc-lib-pdf, but decided to refactor the whole thing to use mPdf so that the contents can be created with a pretty normal phtml template. Much easier to manage and customize now, I believe.

demiankatz commented on languages/Payment/en.ini:
```diff
@@ -0,0 +1,41 @@
+Date = "Date"
+Details = "Details"
+download_receipt_pdf = "Download Receipt (PDF)"
+email_address_missing = "Email address is missing"
```
Might this make sense as a top-level string, since it could apply in multiple contexts? I don't see a reason for specifically scoping this to the payment context (though it seems justifiable for all the other lines in the file).

demiankatz commented on module/VuFind/src/VuFind/Controller/Feature/OnlinePaymentTrait.php:
```diff
@@ -0,0 +1,329 @@
+                'Content-disposition: inline; filename="' .
+                addcslashes($data['filename'], '"') . '"'
+            );
+            echo $data['pdf'];
```
I generally try to avoid hard-coded `exit` calls in controllers, preferring to leverage the response object, redirect helper, etc. Maybe all of that is off the table with the impending exit of laminas-mvc, but I still wonder if we should try to be more formal, or at least encapsulate some of this logic outside of controller-land.

EreMaijala replied:
Good point. Now revised.

demiankatz commented on module/VuFind/src/VuFind/Controller/Feature/OnlinePaymentTrait.php:
```diff
@@ -0,0 +1,329 @@
+            // We should only end up here on error, but redirect always just in case
+            // the payment handler somehow misbehaves:
+            header('Location: ' . $this->getServerUrl('myresearch-fines'));
+            exit();
```
If this function almost always exits, it does feel to me like returning something the calling code could use might feel cleaner...

demiankatz reviewed (changes requested):
Thanks, @EreMaijala -- I've only had time so far to look at the first 24 files in this PR, but here are my comments so far. I'll look at another chunk when time permits!

EreMaijala committed:
Move email_address_missing.

EreMaijala committed:
Return a Response from handleOnlinePayment as needed

EreMaijala committed:
Fix code style and receipt.

EreMaijala committed:
Add test for receipt link.

EreMaijala committed:
Fix code style.

demiankatz commented on module/VuFindDevTools/src/VuFindDevTools/Controller/PaymentServiceController.php:
```diff
@@ -0,0 +1,215 @@
+                    $returnUrl = $this->addSignature($session['returnUrl']);
+                    header("Location: $returnUrl", true, 302);
+                }
+                exit();
```
Can we do this with the framework instead of hard-coded exit?

demiankatz commented on module/VuFind/src/VuFind/ILS/Driver/MultiBackend.php:
```diff
@@ -1119,14 +1142,14 @@ protected function getSource($id)
+            return $this->getSourceFromParams($params, (array)$checkFields);
         } else {
-            $source = $this->getSourceFromParams($params);
+            return $this->getSourceFromParams($params);
```
You could just get rid of the `else` clause and make this the default return at the end of the function.

demiankatz commented on module/VuFind/tests/integration-tests/src/VuFindTest/Mink/OnlinePaymentTest.php:
```diff
@@ -0,0 +1,710 @@
+     */
+    public static function paymentProvider(): array
+    {
+        return [
```
Is it worth naming these test cases with array keys?

demiankatz commented on module/VuFind/tests/integration-tests/src/VuFindTest/Mink/OnlinePaymentTest.php:
```diff
@@ -0,0 +1,710 @@
+    {
+        $blockMsg = 'You have fees that cannot be paid online. Please contact the library customer service.';
+        return [
+            [
```
Named test cases might be helpful here too.

demiankatz reviewed (changes requested):
Chipping away at this review gradually -- see below for comments from looking at tests and devtools.

EreMaijala committed:
Disable parts of test that don't work with chrome-headless-shell.

EreMaijala committed:
Use framework in PaymentServiceController.

EreMaijala committed:
Clean up MultiBackend.

EreMaijala committed:
Add names to tests.

EreMaijala commented:
@demiankatz Latest comments addressed. Unfortunately I also noticed that the latest added test for receipt PDF doesn't work properly with chrome-headless-shell (I had used visible Chrome when working on the test), so I commented it out. I *does* work with headless Chrome though, so perhaps it can be re-enabled in future.

EreMaijala committed:
Skip tests if not in correct mode.

EreMaijala committed:
Fix code style.

demiankatz commented on module/VuFind/tests/integration-tests/src/VuFindTest/Mink/OnlinePaymentTest.php:
```diff
@@ -76,6 +76,22 @@ public static function setUpBeforeClass(): void
+        parent::setUp();
+
+        // Skip tests if we're not running in development or testing mode:
+        if (!in_array(APPLICATION_ENV, ['development', 'testing'])) {
```
Does this actually work? Can we really assume that the Mink tests are running in the same environment as the code being tested? My proposal was to expose the level through a meta tag and then read it in the test to be sure. If this actually works, I'd of course be happier to keep things simple -- but I didn't think the tests had a way of knowing the environment, since what's set in Apache may be different from what's exported at the command line.

EreMaijala replied:
Oops, I totally failed to implement it properly. After a bit of thought I didn't feel comfortable exposing the mode in a meta tag, so I added a check that blocks payment when Test handler is used and VuFindDevTools module isn't loaded.

demiankatz commented on module/VuFind/src/VuFind/Db/Entity/Payment.php:
```diff
@@ -0,0 +1,713 @@
+     */
+    public function getStatus(): PaymentStatus
+    {
+        return PaymentStatus::from($this->status);
```
Since we discussed the limitations of enums elsewhere -- is there any possibility that we might need to support custom values here? (I imagine there's a good chance that this is fine as-is; just asking the question before we lock it in).

EreMaijala replied:
I think that this is a case where custom values should not be supported. Any new ones need to be carefully considered and their behavior defined (e.g. is it resolvable or does it block payment).

demiankatz commented on module/VuFind/src/VuFind/Db/Entity/Payment.php:
```diff
@@ -0,0 +1,713 @@
+     *
+     * @return static
+     */
+    public function setCanceled(): static
```
I wonder if these methods should be named differently -- they're not traditional setters, so calling them setX could be misleading; maybe something like `applyCanceledStatus` and `applyFailureStatus` would better convey that they're setting multiple fields, etc.

EreMaijala replied:
Now renamed.

demiankatz commented on module/VuFind/src/VuFind/Db/Entity/Payment.php:
```diff
@@ -0,0 +1,713 @@
+    /**
+     * Set status message.
+     *
+     * Note that some other methods override the status message, so ensure that this is called last if required!
```
Do we need a similar note on setStatus above? Seems like the same situation.

demiankatz commented on module/VuFind/src/VuFind/Db/Entity/Payment.php:
```diff
@@ -0,0 +1,713 @@
+    public function isRegistrationInProgress(): bool
+    {
+        $startDate = $this->getRegistrationStartDate();
+        return $startDate && (time() - $startDate->getTimestamp() < 120);
```
Why 120 seconds specifically? Should this be configurable or at least a constant?

EreMaijala replied:
Well, it was just based on the idea that any ILS driver should be able to respond in 60 seconds, so double that to be super-safe. I moved it to a constant.

demiankatz reviewed (changes requested):
Thanks, @EreMaijala -- still struggling to get caught up, but I looked at your most recent changes and the new Payment entity tonight.

EreMaijala committed:
Skip tests if VuFindDevTools is not loaded.

EreMaijala committed:
Drop unused method.

EreMaijala committed:
Fix code style.

EreMaijala committed:
Rename Payment entity methods.

EreMaijala committed:
Improve VAT percent support.

EreMaijala committed:
Refactor receipt PDF to use HTML template and mpdf.

Also includes improved VAT breakdown table.

EreMaijala committed:
Fix composer dependencies.

EreMaijala committed:
Remove KohaRest::getOnlinePaymentDetails to use the trait version

demiankatz commented on module/VuFind/src/VuFind/OnlinePayment/OnlinePaymentManager.php:
```diff
@@ -0,0 +1,676 @@
+            }
+            // Check that DevTools module is available when using the Test handler:
+            if ($this->getHandlerName($sourceIls) === 'Test') {
+                global $modules;
```
Using a global doesn't really appeal to me. A couple of possible alternatives:

1.) Check for `APPLICATION_ENV === 'development' || APPLICATION_ENV === 'testing'` -- it's duplicated logic, which isn't great, but it feels better to me than relying on a global variable.

1a.) A variation on the above: check for `APPLICATION_ENV !== 'production'`. It makes sense to disable a test handler in a production environment regardless of the implementation details of how the DevTools module is involved in the process. Maybe we should just rely on that fact to keep things simple.

2.) Define a new constant in application.config.php, like `VUFIND_DEVTOOLS_LOADED`. I realize that's essentially just a global variable by another name, but it feels a little more intentional and robust.

I tend to lean toward 1a as the best solution, but happy to discuss further!

EreMaijala replied:
I almost used 1a, but there could be situations where someone would like to test payment with ILS in otherwise production system. I ended up adding Module Manager as dependency as asking it. Do you find this acceptable?

demiankatz replied:
How about instead of making ModuleManager a dependency, you add a parameter or option to the constructor to enable/disable test mode, and then you use ModuleManager in the factory to set the switch? I'm concerned that ModuleManager is going to get deprecated soon, so I don't want to embed that too deeply in our code, but if we put that logic in the factory, it's fine for now and we can come up with a different detection method in the future when we are forced to remove ModuleManager (without having to worry about a breaking constructor change).

EreMaijala replied:
Good idea, done.

demiankatz commented on module/VuFind/src/VuFind/Db/Entity/PaymentFee.php:
```diff
@@ -0,0 +1,370 @@
+    protected $payment;
+
+    /**
+     * Title
```
This is the only property missing a period for some reason.

```suggestion
     * Title.
```

demiankatz commented on module/VuFind/src/VuFind/Db/Entity/PaymentFee.php:
```diff
@@ -0,0 +1,370 @@
+    }
+
+    /**
+     * Get tax percent (in 1/100ths or a percent).
```
Typo:

```suggestion
     * Get tax percent (in 1/100ths of a percent).
```

...and note that this "or a percent" typo is repeated in multiple places, so you might want to do a search-and-replace to fix them all.

EreMaijala replied:
Oops, fixed.

demiankatz commented on module/VuFind/src/VuFind/Db/Entity/PaymentFee.php:
```diff
@@ -0,0 +1,370 @@
+     */
+    public function calculateAmountExcludingTax(): int
+    {
+        return (int)round($this->amount / (1 + $this->getTaxPercent()));
```
I don't think I understand this math. If I spend $1.00 with a 6% tax,, my total is going to be $1.06. If I'm understanding this class correctly, that means that $this->amount should be 106 (106 pennies), and $this->getTaxPercent() should return 600 (since 6% = 600 hundredths of a percent). The amount excluding tax should be the original $1.00, but 106 / (1 + 600) = 106 / 601 = 0.176, which rounds down to zero. What am I missing/misunderstanding?

EreMaijala replied:
It was indeed very much broken with the divider for the percentage missing. Should be better now. The truth is that we currently only deal with fees with no VAT applied, so the support for taxes is new and hasn't been tested.

demiankatz replied:
Seems like a sign that we at least need some unit tests on this part of the code!

EreMaijala replied:
True, I'll see if I can come up with something. :)

EreMaijala replied:
I added a simple test that checks the VAT breakdown on the receipt.

demiankatz commented on module/VuFind/src/VuFind/Db/Service/PaymentFeeService.php:
```diff
@@ -0,0 +1,91 @@
+     *
+     * @param PaymentEntityInterface $payment Payment
+     *
+     * @return array
```
Would it be safe to say `string[]` here, or are the types inside the array potentially mixed?

demiankatz commented on module/VuFind/src/VuFind/Db/Service/PaymentService.php:
```diff
@@ -0,0 +1,363 @@
+use function intval;
+
+/**
+ * Database service for payment transactions.
```
Class comment doesn't match file comment; intentional?

EreMaijala replied:
No, fixed.

demiankatz commented on module/VuFind/src/VuFind/Db/Service/PaymentService.php:
```diff
@@ -0,0 +1,363 @@
+    public function createEntity(): PaymentEntityInterface
+    {
+        $entity = $this->entityPluginManager->get(PaymentEntityInterface::class);
+        $entity->setCreated(new DateTime());
```
It seems a little unusual to have the `createEntity` method do things differently from the entity object's constructor. Would it be better to have a `createInProgressPayment` method to do the extra initialization, and keep `createEntity` generic?

demiankatz commented on module/VuFind/src/VuFind/Db/Service/PaymentService.php:
```diff
@@ -0,0 +1,363 @@
+                . 'FROM ' . PaymentEntityInterface::class . ' p '
+                . 'WHERE p.id = :id';
+        $parameters = compact('id');
+        $query = $this->entityManager->createQuery($dql);
```
Doesn't the entityManager have a method to look up an entity by ID? That might allow simplification of this method if I'm remembering correctly.

EreMaijala replied:
Indeed it does, and now it's used. We seem to have several other places where `$this->entityManager->find()` could be used as well.

demiankatz replied:
Yeah, I think I learned about it somewhat late in the process. We should go back and clean up those other ones when time permits, but for now I'm just trying not to add more. :-)

demiankatz commented on module/VuFind/src/VuFind/Db/Service/PaymentService.php:
```diff
@@ -0,0 +1,363 @@
+        ];
+
+        $dql = 'SELECT p FROM ' . PaymentEntityInterface::class . ' p'
+            . ' WHERE p.catUsername = :catUsername AND p.status IN (' . implode(',', $statuses) . ')'
```
Do you need to implode the statuses here? Can you use a placeholder and pass the array as a parameter to take better advantage of built-in escaping?

demiankatz commented on module/VuFind/src/VuFind/Db/Service/PaymentService.php:
```diff
@@ -0,0 +1,363 @@
+        ];
+
+        $dql = 'SELECT p FROM ' . PaymentEntityInterface::class . ' p'
+            . ' WHERE p.catUsername = :catUsername AND p.status IN (' . implode(',', $statuses) . ')'
```
Same question here.

demiankatz commented on module/VuFind/src/VuFind/Db/Service/PaymentService.php:
```diff
@@ -0,0 +1,363 @@
+     */
+    public function getUnresolvedPaymentsToReport(int $interval): array
+    {
+        $statuses = implode(',', [PaymentStatus::FinesUpdated->value, PaymentStatus::RegistrationExpired->value]);
```
Here too. If this can be simplified, maybe worth searching the code for implode in case I missed any others, and so I don't have to keep repeating this comment. :-)

demiankatz commented on module/VuFind/src/VuFind/Db/Service/PaymentService.php:
```diff
@@ -0,0 +1,363 @@
+     *
+     * @return void
+     */
+    public function refreshEntity(PaymentEntityInterface $entity): void
```
Is this intentionally missing from PaymentServiceInterface?

EreMaijala replied:
It wasn't missing, the methods just weren't in the same order. Order fixed.

demiankatz reviewed (changes requested):
I've made it past the halfway mark in reviewing files here -- posting another partial review, focused mostly on recent changes and the database services.

demiankatz commented on module/VuFind/src/VuFind/ILS/Driver/Demo.php:
```diff
@@ -1167,6 +1175,62 @@ public function getMyFines($patron)
+                    $paid += $fine['balance'];
+                }
+            }
+        }
```
Shave off some lines?

```suggestion
        foreach ($session->fines ?? [] as $key => $fine) {
            if (
                ($fine['payable_online'] ?? false)
                && (!$fineIds || in_array($fine['fine_id'] ?? '', $fineIds))
            ) {
                unset($session->fines[$key]);
                $paid += $fine['balance'];
            }
        }
```

demiankatz commented on module/VuFind/src/VuFind/ILS/Logic/OnlinePaymentTrait.php:
```diff
@@ -0,0 +1,140 @@
+     */
+    protected function fineBlocksPayment(array $fine): bool
+    {
+        $paymentConfig = $this->config['OnlinePayment'] ?? [];
```
Do we need a comment somewhere indicating that the `$this->config` property is expected (or at least might be checked)? Do we need to clarify in the header comments that this is intended for inclusion in an ILS driver?

demiankatz commented on module/VuFind/src/VuFind/ILS/Logic/OnlinePaymentTrait.php:
```diff
@@ -0,0 +1,140 @@
+            return true;
+        }
+        foreach ((array)($paymentConfig['blockingNonPayableDescriptions'] ?? []) as $pattern) {
+            if (str_starts_with($pattern, '/') && str_ends_with($pattern, '/')) {
```
Do we need to document that only `/.../` regex patterns are supported? Do we need to support modifiers?

EreMaijala replied:
Support for modifiers added and comments in .ini files improved.

demiankatz reviewed (changes requested):
A little more review progress, this time focused on ILS drivers.

EreMaijala committed:
Use ModuleManager to check for VuFindDevTools module.

EreMaijala committed:
Update module/VuFind/src/VuFind/Db/Entity/PaymentFee.php

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

EreMaijala committed:
Fix typos and tax-less amoutn calculation.

EreMaijala committed:
improve comments.

EreMaijala committed:
Update module/VuFind/src/VuFind/ILS/Driver/Demo.php

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

EreMaijala committed:
Address review comments.

EreMaijala committed:
Rename payment creation method.

EreMaijala committed:
Restore createEntity.

EreMaijala committed:
Merge branch 'dev' into dev-online-payment-support2

EreMaijala committed:
Check devtools in factory. Fix year.

EreMaijala committed:
Rename variable.

EreMaijala committed:
Fix code style.

demiankatz commented on module/VuFind/src/VuFind/ILS/Logic/OnlinePaymentTrait.php:
```diff
@@ -0,0 +1,144 @@
+            return true;
+        }
+        foreach ((array)($paymentConfig['blockingNonPayableDescriptions'] ?? []) as $pattern) {
+            if (str_starts_with($pattern, '/') && preg_match('{/\w?$}', $pattern)) {
```
If we're using a regex anyway, would it be just as simple to:

```suggestion
            if (preg_match('{^/.*/\w?$}', $pattern)) {
```

EreMaijala replied:
Sure, but if the pattern doesn't start with a slash, we can avoid compiling a regex. I know the perf difference is negligible, but I still like to avoid running a regex if it's not required.

demiankatz replied:
Fair enough!

demiankatz reviewed (changes requested):
Thanks for all the improvements. Just one new comment on the latest batch of changes. I'll try to look at some more files in a little bit!

sturkel89 commented:
@EreMaijala - I'm curious to test this PR, but I can't figure out how to get the test API key from Stripe. Once I have it, I think I'll put it here in Demo.ini:

; *** Stripe ***
handler = Stripe
apiKey = "sk_test..."

Can you explain? Thanks! :)

demiankatz commented on module/VuFind/src/VuFind/OnlinePayment/Handler/AbstractBase.php:
```diff
@@ -0,0 +1,550 @@
+    {
+        header("Location: $url", true, 302);
+        $this->addPaymentEvent($payment, AuditEventSubtype::Payment, 'Redirected to payment gateway');
+        exit();
```
Here's another exit statement; can we avoid this?

demiankatz commented on module/VuFind/src/VuFind/OnlinePayment/Handler/AbstractBase.php:
```diff
@@ -0,0 +1,550 @@
+        } else {
+            // We don't have both names separately, try to extract first name from
+            // last name.
+            if (strpos($lastname, ',') > 0) {
```
Do you need to worry about suffixes; I imagine this code wouldn't handle "John Smith, Jr." correctly, for example.

EreMaijala replied:
That was never a great implementation, and it's probably better to not even try, so I eliminated the whole method.

demiankatz commented on module/VuFind/src/VuFind/OnlinePayment/Handler/PluginManager.php:
```diff
@@ -0,0 +1,77 @@
+     * @var array
+     */
+    protected $aliases = [
+        'Paytrail' => Paytrail::class,
```
I feel like we usually use lowercase aliases; is there a reason to change the pattern here?

EreMaijala replied:
No, just an oversight. Fixed.

demiankatz commented on module/VuFind/src/VuFind/OnlinePayment/OnlinePaymentEventTrait.php:
```diff
@@ -0,0 +1,82 @@
+
+namespace VuFind\OnlinePayment;
+
+use VuFind\Controller\AbstractBase;
```
Maybe importing this `as AbstractController` or something would make the code using it easier to understand.

demiankatz commented on module/VuFind/src/VuFind/OnlinePayment/OnlinePaymentHttpRequestTrait.php:
```diff
@@ -0,0 +1,212 @@
+ *
+ * PHP version 8
+ *
+ * Copyright (C) The National Library of Finland 2016-2023.
```
Use this year?

```suggestion
 * Copyright (C) The National Library of Finland 2016-2025.
```

demiankatz commented on module/VuFind/src/VuFind/OnlinePayment/OnlinePaymentManager.php:
```diff
@@ -0,0 +1,675 @@
+     * @throws ILSException
+     * @return array Associative array of payment details
+     *
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
```
It looks like you use all of the parameters...

demiankatz commented on themes/bootstrap5/templates/OnlinePayment/receipt.phtml:
```diff
@@ -0,0 +1,219 @@
+<?php
+// Note: this template is used to create a payment receipt as a complete HTML page. The result is passed to mpdf for PDF
+// conversion. Not all HTML features are supported, and external stylesheets cannot not be used.
```
Double-negative:

```suggestion
// conversion. Not all HTML features are supported, and external stylesheets cannot be used.
```

demiankatz commented on themes/bootstrap5/templates/OnlinePayment/receipt.phtml:
```diff
@@ -0,0 +1,219 @@
+  <head>
+    <title><?=$titleHtml?> - <?=$this->payment->getPaidDate()->format('Y-m-d H-i')?></title>
+    <meta name="description" content="<?=$this->transEscAttr('Payment::breakdown_title')?>">
+    <style>
```
I realize you can't use an external stylesheet, but maybe it would make sense to break out this `<style>` tag into an included sub-template for easier customization (without having to copy the whole template).

demiankatz reviewed (changes requested):
Okay, another incremental review -- I've made my way through the OnlinePayment namespace. It's a LOT of code, and I don't claim to have read it all thoroughly, but see below for a few things I noticed along the way.

EreMaijala committed:
Update module/VuFind/src/VuFind/OnlinePayment/OnlinePaymentHttpRequestTrait.php

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

EreMaijala commented:
> @EreMaijala - I'm curious to test this PR, but I can't figure out how to get the test API key from Stripe. Once I have it, I think I'll put it here in Demo.ini:
> 
> ; *** Stripe *** handler = Stripe apiKey = "sk_test..."
> 
> Can you explain? Thanks! :)

You'll need to create a Stripe account first (it's free). I didn't manage to log in using Firefox, but Chrome seems to work. When logged in, ensure that you're in Sandbox mode (top of the screen). Then in the bottom left there a menu entry "Developers". In the popup menu you'll find "API keys".
Direct link: https://dashboard.stripe.com/test/apikeys
Docs that include test card numbers: https://docs.stripe.com/testing

EreMaijala committed:
Update themes/bootstrap5/templates/OnlinePayment/receipt.phtml

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

EreMaijala committed:
Fix fines table, move tax rate mapping to driver, add test for receipt with tax rate, unify getMyFines array keys.

EreMaijala committed:
Fix event logging error in Paytrail handler.

EreMaijala committed:
Return Response instead of exit.

EreMaijala committed:
Lowercase aliases.

EreMaijala committed:
Cleanup.

EreMaijala committed:
Split style to a separate template.

EreMaijala committed:
Eliminate extractUserNames.

demiankatz commented on module/VuFind/tests/integration-tests/src/VuFindTest/Mink/OnlinePaymentTest.php:
```diff
@@ -0,0 +1,763 @@
+            ],
+            [
+                true,
+            ],
```
Label and condense?

```suggestion
            'without multibackend' => [false],
            'with multibackend' => [true],
```

demiankatz commented on module/VuFind/tests/integration-tests/src/VuFindTest/Mink/OnlinePaymentTest.php:
```diff
@@ -0,0 +1,763 @@
+        /*
+
+        TODO: This isn't working properly with chrome-headless-shell but results in "Page not loaded" exception
+        from $session->reload()
```
Might this be caused by lack of PDF support?

EreMaijala replied:
I suppose that's possible. I couldn't think of a meaningful workaround, so I removed that part of the test.

demiankatz commented on module/VuFind/tests/integration-tests/src/VuFindTest/Mink/OnlinePaymentTest.php:
```diff
@@ -0,0 +1,763 @@
+     *
+     * @return array
+     */
+    protected function getFakeFines(string $bibId)
```
Missing return type.

```suggestion
    protected function getFakeFines(string $bibId): array
```

demiankatz commented on module/VuFindAdmin/src/VuFindAdmin/Controller/OnlinePaymentController.php:
```diff
@@ -0,0 +1,249 @@
+
+            $paymentEntity->applyRegistrationResolvedStatus();
+            $paymentService->persistEntity($paymentEntity);
+            $this->addPaymentEvent($paymentEntity, 'Payment marked as resolved');
```
Philosophical question on audit logging: is it enough to log after the event has occurred? What happens if the database breaks between persisting the entity and writing the log? We then have a phantom transaction. I realize the likelihood of this is vanishingly small, but would the audit trail be more reliable if we logged both before and after persistence operations for critical transactions?

(I'm not insisting on, or even suggesting, a change -- but just thinking about edge cases and wondering if we need to worry about highly unlikely but theoretically possible events like this).

EreMaijala replied:
That is a real issue, but more likely to occur when interacting with third-party systems such as the ILS. I've tried to keep a balance between enough logging and trying to avoid logging too much so that it doesn't become noise. I suppose one option would be to log before, do the work, log after and remove the log entry from before, but that would show as a gap in the audit trail.
In any case, if the database connection fails after persisting the payment entity, the log entry about it being done would be lost and the trail would be incomplete. I suppose we could wrap these into a database transaction too so that it's all or nothing. If you think that would be warranted, I can try to add transactions where possible.

demiankatz replied:
Maybe it makes sense to add a `persistEntityWithAuditMessage` method to the OnlinePaymentManager which does the persistance-and-auditing as a transaction, as you propose. Then the OnlinePaymentController could use the OnlinePaymentManager instead of implementing the OnlinePaymentEventTrait locally.

EreMaijala replied:
Makes sense. First I'll just add a trait for transactions in a separate PR. It seems that a lot of the services duplicate the methods. Or move the methods to the base class.

EreMaijala replied:
Now done.

demiankatz commented on module/VuFindConsole/src/VuFindConsole/Command/Util/ConsoleLoggerTrait.php:
```diff
@@ -0,0 +1,210 @@
+ * @link     http://vufind.org/wiki/vufind2:developer_manual Wiki
+ * @todo     Use Symfony output
+ */
+trait ConsoleLoggerTrait
```
Looking at this, I'm a little confused about if/how it relates to normal logging functionality (as impacted by #4429) and/or if it should in some way.

We might want to either align our method names more closely with existing logging methods, or else differentiate them more carefully, depending on our goals....

EreMaijala replied:
This doesn't actually have any relation to normal logging. This got copied from our relatively ancient codebase without further thought, but needs some attention indeed. Perhaps something like ConsoleOutputTrait would make more sense. In any case, the idea here is that you can output different stuff to stdout and stderr so that when run from crontab, you can redirect stdout to a log file but let cron email any errors written to stderr (with any personal details scrubbed) to you.
Though now that I look at it, I think we may not need the trait at all.

EreMaijala replied:
I've now moved the monitor to online_payment/monitor and eliminated the trait.

demiankatz commented on module/VuFindConsole/src/VuFindConsole/Command/Util/ConsoleLoggerTrait.php:
```diff
@@ -0,0 +1,210 @@
+     * Output an error message with a timestamp
+     *
+     * @param string $msg          Message
+     * @param string $publishedMsg Published version of the error message. Must
```
I find this mechanism a bit puzzling. Why the `=` mechanism instead of something like a null default? Why implement it here but not in other methods?

demiankatz commented on module/VuFindConsole/src/VuFindConsole/Command/Util/OnlinePaymentMonitor.php:
```diff
@@ -0,0 +1,356 @@
+#[AsCommand(
+    name: 'util/online_payment_monitor'
+)]
+class OnlinePaymentMonitor extends Command
```
When time permits, it would be good to add a test for this class -- I realize 100% coverage may be challenging, but even a stub to get the work started would be better than nothing.

EreMaijala replied:
Stub added.

demiankatz commented on themes/bootstrap5/js/fines.js:
```diff
@@ -0,0 +1,169 @@
+  /**
+   * Set the aria-live attribute for an element
+   * @param {Element} element Element to set aria live into
+   * @param {string} politeness Aria live value
```
Do you need to clarify that an empty value removes the attribute?

```suggestion
   * @param {string} politeness Aria live value (empty string to remove attribute)
```

demiankatz commented on themes/bootstrap5/js/fines.js:
```diff
@@ -0,0 +1,169 @@
+      .then(() => {
+        // Clear account notification cache and reload current page without parameters
+        VuFind.account.clearCache();
+        location.href = window.location.href.split('?')[0];
```
Would it make sense to refactor the location refresh to a `finally` since it's the same behavior on either branch? (If you think it's more clear as-is, that's fine with me... the repetition just caught my eye).

demiankatz commented on themes/bootstrap5/templates/admin/payment/details-table.phtml:
```diff
@@ -0,0 +1,45 @@
+    <td><?=$this->escapeHtml($paymentEntity->getId())?></td>
+  </tr>
+  <tr>
+    <th><?=$this->transEsc('Local Identifier')?></th>
```
I think some of these labels do not exist in the language files. I realize that we haven't gone to the effort of translating the admin pages, so maybe they don't need to -- but in that case, maybe it's confusing to run them through the translator. If we do want to translate undefined strings, maybe we should at least give them a better text domain; shouldn't these go in Payment:: if we use them at all?

EreMaijala replied:
Translation in admin templates is common occurrence already. And I think that it's a nice gesture for anyone wanting to add a translation (even if just locally for one task).
Many of the strings are pretty generic, and some have existing translations. That's why I used the default domain. But some are admittedly specific to payment, and the division is not easy. I can move them all to Payment domain if that's better.

demiankatz replied:
I think if we're going to have translations, we might as well use the Payment domain -- the specificity is useful for at least some of these values, and it's now easy enough to alias to default translations if necessary, so I think it covers the most potential use cases.

demiankatz reviewed (changes requested):
I have finished my first pass through all the files -- see below for more comments!

After this, I think the next step is to see what @sturkel89 has to say about the functional aspects of the feature (and find out whether she can break anything).

sturkel89 commented:
Thanks for your help! I set up my Stripes account and was able to test the PR (up to a point). 

I was able to see the information about my balance due and what's payable online, as well as the Pay Online button:

<img width="865" height="675" alt="image" src="https://github.com/user-attachments/assets/761d455c-04d9-40ed-b1d0-d51a25659a32" />

However, when I selected some of my fines, clicked the Pay Online button, and chose to proceed with the Stripe payment, I got this error:

<img width="1247" height="815" alt="image" src="https://github.com/user-attachments/assets/065fc03d-6583-49fb-abfa-d5b936a40ada" />

Can you troubleshoot? Thanks! I'll be happy to comment on more aspects of this new functionality once this error is cleared up.

EreMaijala committed:
Update themes/bootstrap5/js/fines.js

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

EreMaijala committed:
Cast fine id to string.

EreMaijala commented:
@sturkel89 Sorry, my bad. The new files use strict typing, and the Demo driver returned an integer instead of string as it should. Should be fixed now!

EreMaijala committed:
Remove the PDF test.

EreMaijala committed:
Eliminate ConsoleLoggerTrait, refactor monitor to online_payment/monitor.

EreMaijala committed:
Remove underscore from command line category.

demiankatz commented on module/VuFindConsole/src/VuFindConsole/Command/OnlinePayment/MonitorCommandFactory.php:
```diff
@@ -0,0 +1,88 @@
+        ?array $options = null
+    ) {
+        // We need to initialize the theme so that the view renderer works:
+        $mainConfig = $container->get(\VuFind\Config\ConfigManager::class)->getConfigObject('config');
+        $theme = new \VuFindTheme\Initializer($mainConfig->Site, $container);
+        $theme->init();
+
+        $dbServiceManager = $container->get(\VuFind\Db\Service\PluginManager::class);
+        return new $requestedName(
+            $dbServiceManager->get(PaymentServiceInterface::class),
+            $container->get(OnlinePaymentManager::class),
+            $container->get('ViewRenderer'),
+            $container->get(Mailer::class),
+            $mainConfig->toArray(),
```
Better to use `getConfigArray` so we don't pack and then unpack the configuration for no reason:

```suggestion
        $mainConfig = $container->get(\VuFind\Config\ConfigManager::class)->getConfigArray('config');
        $theme = new \VuFindTheme\Initializer($mainConfig->Site, $container);
        $theme->init();

        $dbServiceManager = $container->get(\VuFind\Db\Service\PluginManager::class);
        return new $requestedName(
            $dbServiceManager->get(PaymentServiceInterface::class),
            $container->get(OnlinePaymentManager::class),
            $container->get('ViewRenderer'),
            $container->get(Mailer::class),
            $mainConfig,
```

EreMaijala replied:
VuFindTheme\Initializer wants a Config object, so I can't avoid it yet.

demiankatz replied:
Ahh, sorry, missed that additional usage there!

demiankatz reviewed (changes requested):
Looks like you're still working through some of my comments (at least, you thumbs-upped some of my suggestions without actually committing them), but in the meantime, here's one small improvement to recently-changed code.

EreMaijala committed:
Use a finally-block for redirect.

EreMaijala committed:
Update module/VuFind/tests/integration-tests/src/VuFindTest/Mink/OnlinePaymentTest.php

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

EreMaijala committed:
Use Payment namespace.

EreMaijala committed:
Update module/VuFind/tests/integration-tests/src/VuFindTest/Mink/OnlinePaymentTest.php

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

EreMaijala committed:
Specify cache directory for mpdf.

EreMaijala committed:
Fix the getFakeFines fix.

sturkel89 commented:
Thanks for fixing that glitch, @EreMaijala ! I've played with the PR for a while today with no other errors popping up. The only payment method I tested was Stripe; it worked seamlessly!

As you know, though, I always have to find something to criticize. ;-) I'll write up my nitpicky thoughts and requests here. Please forgive me if I request functionality that DOES exist but I couldn't see how to change the parameters.

-----

### Capitalization in the Your Fines display

This is the top of the Your Fines display, above the table:

<img width="823" height="349" alt="image" src="https://github.com/user-attachments/assets/c97fde18-97bf-46f1-88b7-b62da8e97419" />

I am not sure if we have any standards for which text is capitalized in VuFind headings, but I think I'd prefer the summary at the top of the page to be presented in sentence case:

> Last paid: $24.50 09-15-2025 10:43               Download receipt (PDF)
> Total balance due: (2): **$29.00**
> Payable online (2): **$29.00**    Minimum payable amount: **$5.00**
> Remaining balance after payment: **$29.00**
> 
> __ Select all entries
> 
> **Pay Online**

---------------

### Internationalization

I see that the date is given in the "American" order of MM/DD/YYYY. But the time doesn't give AM/PM, so I think it may be 24 hour time. I would prefer that this all be consistent.

Also, the receipt shows information for VAT, which isn't relevant for the U.S. and probably some other countries. Could there be a setting that controls all of these preferences, so that libraries in the U.S. get their most relevant configuration by default, and the other most prevalent VuFind user countries also have a place to select a setting that gives them their most likely-to-be-preferred display?

-----------

### Admin Module

I noticed that the Admin page for Online Payment Management seems to focus on the level of the transaction, not on the specific items that were paid for. Is this deliberate? (Do we assume that the libraries will access that information via their native ILS?) Here's the main admin page:

<img width="1194" height="834" alt="image" src="https://github.com/user-attachments/assets/8841626a-8ae2-4caa-832f-c3568f715b03" />

I think it might be helpful to add the payment amount to that table, just to make it easier for the administrator to identify the transaction without having to get into the non-human-friendly Identifier values.

When I went to the **Details** for any transaction, I got a list of steps like "successfully registered" and "redirected to payment gateway" and "payment marked as paid" but nothing about the amount paid or the items that were involved. From my perspective, that's what I expect to see under Details, but again the actual Circulation Desk staff could probably access all of that information via the native ILS in staff view. 

In the Details view for every one of my transactions, I saw the following error with many lines of code: "Sending of receipt failed" and a lot of references to Symfony\Component\Mailer. Is this expected, due to some email server configuration I didn't change?

<img width="1031" height="627" alt="image" src="https://github.com/user-attachments/assets/f467ea60-fde7-4a7f-91a2-d6214909115b" />

--------------

### Receipts

These are all the settings I could find for receipts. 

<img width="889" height="375" alt="image" src="https://github.com/user-attachments/assets/89432f90-2653-4c25-8575-eaec3620c37a" />

Here are a few example receipts:

<img width="682" height="428" alt="image" src="https://github.com/user-attachments/assets/c74b06dc-e61b-4840-8944-9598f3db3a17" />

<img width="712" height="390" alt="image" src="https://github.com/user-attachments/assets/cfccee64-38c7-4e79-a7b8-7427de46eec8" />

Bottom part only:
<img width="712" height="278" alt="image" src="https://github.com/user-attachments/assets/8abab92a-6c20-4c22-a07e-0a7b94db2f74" />

Here are some suggestions re: the receipt.

- Can "default" be removed from the **Payment Recipient** line?
- Could there be an option to enter custom hyperlink text instead of displaying the full hyperlink in the **Contact Information** area?
- I think the receipt is missing some information that might be useful for the patron, especially if they might need to be reimbursed for these fees. Examples: patron name, payment method, last four digits of payment credit card (if applicable).
- With a long item title, the column headings for **Identifier** and **Type** run together. In some cases, the contents of the **Type** and **Details** columns run together (see last example). Word wrap is invoked for extremely long titles (see first example). Is there any way to force minimum padding for each column, regardless of title length? Receipts where the item titles were short all look great and are very readable.
- If the item had **n/a** in the title in VuFind, can the receipt say the same? Apparently a blank is passed along to Stripe, but in the **My Fines** view in VuFind the n/a is displayed. I don't know if this is a weirdness of the Demo driver, or if items not owned by the home library like ILL titles might also have an n/a by the time the fine is assessed – but in any case, the blank value under **Details** is confusing.

----------------------

Please let me know if there are any specific features you would like me to test. I tried adding a new non-payable type of fine (Long Overdue) and can see that it was successful.

<img width="959" height="210" alt="image" src="https://github.com/user-attachments/assets/a7997cab-6839-4429-9489-2730fe1897ec" />

--------

All in all, this PR adds great functionality to VuFind! Very nice work.

EreMaijala committed:
Merge remote-tracking branch 'origin/dev' into dev-online-payment-support2

EreMaijala committed:
Use transactions.

EreMaijala committed:
Use transaction also in monitor.

EreMaijala committed:
Disable VAT breakdown by default and add config for it.

EreMaijala committed:
Add more information to admin.

EreMaijala committed:
Use sentence case more.

EreMaijala committed:
Improve recipient name and filename handling.

EreMaijala committed:
Improve receipt column padding.

EreMaijala committed:
Add support for link text to contact info.

EreMaijala commented:
@sturkel89 Thanks for the detailed report!

- I tweaked the capitalization. The "Total Balance Due" was an existing one, I hope it's fine to update it as well.
- Time format comes from the displayTimeFormat in config.ini. Something like this would give you 12 hour clock with AM/PM: `displayTimeFormat = "h:i A"`
- I added more details to the Admin module. There's now also a list of fees in the details view.
- If e.g. sending a record by email works for you, then there's a problem with the automatic emailing of the PDF receipt. Otherwise it's just that you'll need to configure VuFind properly so that it can send email.

Receipt:
- "default" should be fixed now.
- I added support for link text in contact info. Not sure if that's useful if the PDF is ever printed, but I suppose it isn't often the case.
- The receipt doesn't contain any patron information because we don't really know why actually did the payment, and I don't want to start tracking too much more personal information. Any reimbursement would be processed in Stripe's or other systems' admin interface anyway, and the details are available there.
- I fixed the column padding, should be more clear now with long titles.
- Fees without a title: Some fees are not even related to titles, so there could be others that don't have any details. Perhaps the column could contain just a dash or something where there's no details available, but "n/a" is a bit problematic to translate in a way that doesn't sound like there's an issue. Luckily an organization can customize the template if necessary. :)

demiankatz commented on module/VuFind/src/VuFind/OnlinePayment/OnlinePaymentManager.php:
```diff
@@ -589,15 +583,63 @@ public function getAndValidateOnlinePaymentConfig(array $patron): array
+
+            $this->addPaymentEvent($payment, AuditEventSubtype::Payment, 'Payment created');
+        } finally {
+            $this->paymentService->commitTransaction();
```
Do we not want to roll back on error?

EreMaijala replied:
Of course we do. Fixed!

demiankatz commented on module/VuFind/src/VuFind/OnlinePayment/OnlinePaymentManager.php:
```diff
@@ -652,6 +694,43 @@ public function getAndClearPaymentSuccessFlag(): bool
+            $this->paymentService->persistEntity($payment);
+            $this->auditEventService->addPaymentEvent($payment, $eventSubtype, $auditMessage, $eventData);
+        } finally {
+            $this->paymentService->commitTransaction();
```
Same question here...

demiankatz reviewed (changes requested):
Thanks, @EreMaijala -- see below for a couple instances of the same question, which may be driven by a lack of deep understanding of how transactions should work on my part. ;-)

Regarding a couple of points in your reply to @sturkel89:

- If we have an existing "Total Balance Due" translation key, I wouldn't want to create work for the translators by making a "Total balance due" key as well. But if you want to keep the existing key and change the English translation, that might be an easy option, as long as it looks appropriate in the existing context(s) where it is used.
- I agree that using a dash for a placeholder for missing titles might be a reasonable approach. Though maybe the best solution of all would be to use a translation placeholder (e.g. `missing_title_placeholder = "-"`), in case there's a better symbol in other languages.

demiankatz commented on module/VuFindConsole/src/VuFindConsole/Command/OnlinePayment/MonitorCommand.php:
```diff
@@ -0,0 +1,378 @@
+ * @link     http://vufind.org/wiki/vufind2:developer_manual Wiki
+ */
+#[AsCommand(
+    name: 'online_payment/monitor'
```
For completeness, we should add this new command somewhere in the console menu, as per #4613. It probably fits in the "Data Management Tools" submenu along with things like scheduledsearch/notify.

EreMaijala replied:
Done! This one is a hard to remember for me..

demiankatz reviewed (changes requested):
One more detail that I nearly forgot!

EreMaijala committed:
Display dash in place of empty description in receipt.

EreMaijala committed:
Rollback on exception.

EreMaijala committed:
Fix console menu and command name.

EreMaijala commented:
@demiankatz I actually already updated the translation for "Total balance due". The only other place where it's used is at the bottom of the fines list, and I think the sentence case looks fine there. I also added the dash for empty details with a translation key.

sturkel89 commented:
It looks like you made all of my fixes! 

Capitalization now looks good. 

Receipts look better! I see the new column padding, and I see that I could put either plain text (like the Circ Desk's phone number) or a URL, with a human-readable hyperlink label. I also see the date and time display settings in config.ini. (I'm still not happy with the leading zeroes in the month and hour, but that may be configurable somewhere and is not related to this PR in any case!) The dashes are now present when there's no title associated with a fee.

Examples: 

<img width="660" height="331" alt="image" src="https://github.com/user-attachments/assets/a69ef6e2-edcd-4757-b066-158b47340293" />

<img width="654" height="325" alt="image" src="https://github.com/user-attachments/assets/837ea84f-c29b-4235-824d-191a5275baf2" />

And the Admin module now has more information on the main screen and on the details screen:

<img width="1191" height="367" alt="image" src="https://github.com/user-attachments/assets/a220160c-60d7-483b-8cb8-ed446958ede2" />

<img width="1104" height="824" alt="image" src="https://github.com/user-attachments/assets/523086f7-fa04-40b3-876c-fc8c9ba82ce4" />

If there are any other features you'd like me to check, I'm happy to do so. As far as I can see right now, it all looks good!

demiankatz reviewed (approved):
I think this is ready to go now! Tests are passing, I don't see any further problems, and @sturkel89's tests have gone well.

I'm going to approve without merging in case you have any last-minute testing requests or concerns, @EreMaijala, but please feel free to hit the merge button when ready.

I've already updated the [changelog](https://vufind.org/wiki/changelog) (see top bullet, as this is probably our single biggest new feature) -- but please feel free to revise if you can improve upon what I've put there.

I'll resolve the JIRA ticket once the PR is merged (or you're welcome to, if you prefer).

Thanks again for contributing this!

EreMaijala commented:
Thanks @sturkel89 and @demiankatz for the huge task of reviewing and testing all this!

I've no further concerns at the moment, but I'm committed to providing any fixes or tweaks necessary in case something arises when other organizations start implementing online payment. I'll merge shortly. :)

stweil commented on module/VuFind/src/VuFind/ILS/Connection.php:
```diff
@@ -857,6 +857,22 @@ protected function checkMethodpatronLogin($functionConfig, $params)
+     *
+     * @return boolean
+     */
+    protected function checkMethodregisterPayment($functionConfig, $params)
```
@EreMaijala, it looks like `checkMethodregisterPayment` is currently unused. I noticed this because it can return an array although its documentation says that it must return a boolean.

demiankatz replied:
@stweil, it's also possible that the documentation is wrong and needs to be corrected. Either way, @EreMaijala, could you open a release-11.0 PR to correct the problem when you have a chance?

stweil replied:
... or maybe remove the function unless there are plans to use it in the future?

EreMaijala replied:
It's definitely used, it's just the docs that are wrong.

demiankatz replied:
(Note that these "check" functions are called dynamically via a naming convention -- that is probably why it is not obvious that the function is being used).

EreMaijala replied:
@demiankatz #4979 fixes the doc comment.
